dragonfly/actions/action_text.py

```diff
@@ -3,109 +3,178 @@
+
+    Arguments:
+     - *spec* (*str*) -- the text to type
+     - *static* (boolean) --
+       if *True*, do not dynamically interpret *spec*
+       when executing this action
+     - *pause* (*float*) --
+       the time to pause between each keystroke, given
+       in seconds
+     - *autofmt* (boolean) --
+       if *True*, attempt to format the text with correct
+       spacing and capitalization.  This is done by first mimicking
+       a word recognition and then analyzing its spacing and
+       capitalization and applying the same formatting to the text.
     """
 
     _pause_default = 0.02
     _keyboard = Keyboard()
     _specials = {
-                 "\n":   typeables["enter"],
-                 "\t":   typeables["tab"],
+                 "\n": typeables["enter"],
+                 "\t": typeables["tab"],
                 }
 
     def __init__(self, spec=None, static=False, pause=_pause_default,
-                 autofmt=False):
+                 autofmt=False, use_hardware=False):
         self._pause = pause
         self._autofmt = autofmt
+        self._use_hardware = use_hardware
         DynStrActionBase.__init__(self, spec=spec, static=static)
 
     def _parse_spec(self, spec):
-        """ Convert the given *spec* to keyboard events. """
+        """Convert the given *spec* to keyboard events."""
+        from struct import unpack
         events = []
-        for character in spec:
-            if character in self._specials:
-                typeable = self._specials[character]
-            else:
-                try:
-                    typeable = Keyboard.get_typeable(character)
-                except ValueError as e:
-                    raise ActionError("Keyboard interface cannot type this"
-                                      " character: %r (in %r)"
-                                      % (character, spec))
-            events.extend(typeable.events(self._pause))
+        if self._use_hardware or require_hardware_emulation():
+            for character in spec:
+                if character in self._specials:
+                    typeable = self._specials[character]
+                    events.extend(typeable.events(self._pause))
+                else:
+                    try:
+                        typeable = Keyboard.get_typeable(character)
+                        events.extend(typeable.events(self._pause))
+                    except ValueError:
+                        raise ActionError("Keyboard interface cannot type this"
+                                          " character: %r (in %r)"
+                                          % (character, spec))
+        else:
+            for character in text_type(spec):
+                if character in self._specials:
+                    typeable = self._specials[character]
+                    events.extend(typeable.events(self._pause))
+                else:
+                    byte_stream = character.encode("utf-16-le")
```

Comment:
A comment would be helpful here, e.g. "The SendInput API only accepts 16-bit scancodes, so we have to break unicode characters above 0xFFFF into two 16-bit scancodes."

When I first read this I was thinking "why not just use ord()?".

Reply:
I could definitely have been more explicit about this. When using this specific keyboard mode, Windows is expecting a UTF-16 little endian byte stream encoded with the proper surrogate pair to be sent as the scancode(s).